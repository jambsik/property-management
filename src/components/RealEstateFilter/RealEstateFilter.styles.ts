export const realEstateFilterRootStyles = {
    margin: '46px',
};
