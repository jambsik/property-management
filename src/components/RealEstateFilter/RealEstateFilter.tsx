import { Box } from '@mui/system';
import React, { ReactNode, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { FilterTypes } from '../../constants/FilterTypes';
import FilterBox from '../../features/FilterBox/FilterBox';
import Input, { InputProps } from '../../features/Input/Input';
import YearPicker, { YearPickerValue } from '../../features/YearPicker/YearPicker';
import { FilterParams } from '../../helpers/simulateBe';
import { ResponseFilter } from '../../models/Response';
import { realEstateFilterRootStyles } from './RealEstateFilter.styles';

interface RealEstateFilterProps {
    filters: Array<ResponseFilter>;
    onSearch: (filters: FilterParams) => void;
}

const RealEstateFilter = ({ filters, onSearch }: RealEstateFilterProps) => {
    const { t } = useTranslation();
    const [filterValues, setFilterValues] = useState<FilterParams>({});

    const getFilterComponent = (type: FilterTypes, componentProps: InputProps): ReactNode => {
        switch (type) {
            case FilterTypes.String:
            case FilterTypes.Number:
                return <Input {...componentProps} type={type} />;
            case FilterTypes.Year:
                return (
                    <YearPicker
                        label={componentProps.label}
                        onChange={(event: React.ChangeEvent<{}>, yearPickerValue: YearPickerValue) => {
                            const hasValue = yearPickerValue?.label;
                            if (hasValue) {
                                setFilterValues({
                                    ...filterValues,
                                    [componentProps.name]: parseInt(yearPickerValue?.label),
                                });
                            }
                        }}
                    />
                );
            default:
                return <Input {...componentProps} type={FilterTypes.String} />;
        }
    };

    const onSearchHandler = () => {
        onSearch(filterValues);
    };

    return (
        <Box sx={realEstateFilterRootStyles}>
            <FilterBox onSearch={onSearchHandler}>
                {filters.map(({ name, type }: ResponseFilter) => (
                    <React.Fragment key={name}>
                        {getFilterComponent(type, {
                            name,
                            label: t(`realEstate.filters.${name}`),
                            type,
                        })}
                    </React.Fragment>
                ))}
            </FilterBox>
        </Box>
    );
};
export default RealEstateFilter;
