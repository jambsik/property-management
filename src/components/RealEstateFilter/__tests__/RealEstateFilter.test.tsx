import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import RealEstateFilter from '../RealEstateFilter';
import { basicFiltersDataMock } from '../../../__tests__/__fixtures__/filtersDataMock';
import { useTranslation } from 'react-i18next';

describe('RealEstate Filter component test', () => {
    const onSearchMock = jest.fn();
    const addFilterMock = jest.fn();
    const { t } = useTranslation();

    const generateInputSelector = (id: string) => `input[id="${id}"]`;

    const pageObject = {
        searcheableButton: 'filter-box-button',
        street: t('realEstate.filters.street'),
        yearSold: t('realEstate.filters.yrsold'),
        salePrice: t('realEstate.filters.saleprice'),
        yearBuilt: t('realEstate.filters.yearbuilt'),
    };

    it('Should render component without errors', () => {
        const { container } = render(<RealEstateFilter filters={basicFiltersDataMock} onSearch={onSearchMock} addFilter={addFilterMock} />);

        expect(onSearchMock).toHaveBeenCalledTimes(0);
        expect(addFilterMock).toHaveBeenCalledTimes(0);

        expect(container).toMatchSnapshot();
    });

    it('Should render filters with properly component', () => {
        const { container } = render(<RealEstateFilter filters={basicFiltersDataMock} onSearch={onSearchMock} addFilter={addFilterMock} />);

        const streetInput = container.querySelector(generateInputSelector('street'));
        const yeearBuiltElement = container.querySelector(generateInputSelector(`year-picker-${pageObject.yearBuilt}`));

        expect(streetInput).toBeDefined();
        expect(yeearBuiltElement).toBeDefined();

        expect(streetInput).toHaveAttribute('type', 'string');
        expect(yeearBuiltElement).toHaveAttribute('type', 'text');
    });

    it('Should dispatch addFilter when value change', async () => {
        const { container, getByText } = render(<RealEstateFilter filters={basicFiltersDataMock} onSearch={onSearchMock} addFilter={addFilterMock} />);

        const streetInput = container.querySelector(generateInputSelector('street'));
        const yeearBuiltElement = container.querySelector(generateInputSelector(`year-picker-${pageObject.yearBuilt}`));

        expect(streetInput).toBeInTheDocument();

        fireEvent.change(streetInput!, {
            target: {
                value: 200,
            },
        });

        expect(addFilterMock).toHaveBeenCalledTimes(1);
        expect(addFilterMock).toHaveBeenCalledWith('street', '200');

        jest.clearAllMocks();

        expect(yeearBuiltElement).toBeInTheDocument();

        fireEvent.change(yeearBuiltElement!, {
            target: {
                value: 1991,
            },
        });

        await waitFor(() => {
            const option = getByText(1991);

            expect(option).toBeInTheDocument();

            fireEvent.click(option);
        });

        expect(addFilterMock).toHaveBeenCalledTimes(1);
        expect(addFilterMock).toHaveBeenCalledWith('yearbuilt', 1991);
    });
});
