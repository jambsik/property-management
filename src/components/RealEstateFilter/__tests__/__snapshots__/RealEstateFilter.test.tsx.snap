// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RealEstate Filter component test Should render component without errors 1`] = `
<div>
  <div
    class="MuiBox-root css-zti9x5"
  >
    <div
      class="MuiBox-root css-1i6msmp"
    >
      <div
        class="MuiBox-root css-iryi8b"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1954myt-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorInfo css-ahcq22-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="street"
            id="street-label"
          >
            realEstate.filters.street
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorInfo MuiInputBase-formControl css-ovthwe-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="street"
              type="string"
              value=""
            />
          </div>
        </div>
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-mmax81-MuiAutocomplete-root"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-jemhih-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorInfo css-ahcq22-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="year-picker-realEstate.filters.yearbuilt"
              id="year-picker-realEstate.filters.yearbuilt-label"
            >
              realEstate.filters.yearbuilt
            </label>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorInfo MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-194ngh9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id="year-picker-realEstate.filters.yearbuilt"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-mmax81-MuiAutocomplete-root"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-jemhih-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorInfo css-ahcq22-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="year-picker-realEstate.filters.yrsold"
              id="year-picker-realEstate.filters.yrsold-label"
            >
              realEstate.filters.yrsold
            </label>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorInfo MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-194ngh9-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id="year-picker-realEstate.filters.yrsold"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1954myt-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorInfo css-ahcq22-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="saleprice"
            id="saleprice-label"
          >
            realEstate.filters.saleprice
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorInfo MuiInputBase-formControl css-ovthwe-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="saleprice"
              type="number"
              value=""
            />
          </div>
        </div>
        <button
          aria-label="searcheable real estate"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
          data-testid="filter-box-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="SearchIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
